### DISCUSSION TITLE
allow for vectorized outputs?

### DISCUSSION THREAD

**Discussion by ftalbrecht** [pymor/pymor#1214] [Ideas]
We (@TizianWenzel) are using pyMORs models to generate input for machine learning algorithms to learn the input-output relationship (think about mapping a `mu` to a `rom.output(mu)`). Depending on the implementation, after training, the resulting networks can often predict the outputs for several parameters at once in a vectorized manner (as for instance `torch`).
Given the current model interface, we need to (artificially) treat each parameter instance separately and use a Python-based `for`-loop or list comprehension.

While it might happen that the time spent in the loop dominates the prediction time (which is not too great an issue), it mainly feels somewhat unnatural to not be able to call
```python
predictions = rom.output(parameter_space.sample_randomly(1000))
```
in these circumstances.

Similar considerations (allowing for vectorized computations for certain backends) lead us to introduce `VectorArrays`, but I am unsure if the concept would carry over here, from a practical standpoint, in particular since time-dependent outputs for a single `mu` are already a `VectorArray`.

I did not ponder this in detail, thoughts:
- distinguishing between an array of numbers which are interpreted as a single `Mu` (as an argument to `model.output`) and an array of `mu`s could be cumbersome
- for the instationary case, say with `N` timesteps, the `model.output(mu)` is already a (usually scalar) `VectorArray` of length `N`.
  For a `list_of_mus` of length `M`, `model.output(list_of_mus)` would have to produce an array of `shape` `M x N`. So would we have different return types or vectorarrays of different dims?
- would `model.solve(list_of_mus)` also make sense? (I do not think so)

**Comment by ftalbrecht** [comment]
Another point:

* we would also need a vectorized representation of parameters, say a numpy array of serialized mus in each row.

**Comment by sdrave** [comment]
Sorry, I somehow missed this thread. So, IMHO, it is not a big issue to write:

```python
predictions = [rom.output(mu) for mu in parameter_space.sample_randomly(1000)]
```

I would not go through the trouble of extending the `Model` interface (and maybe even the `Operator` interface as well), just to save a few characters. Regarding runtime considerations: If your ROM is so fast that the Python loop matters, then you probably want a custom implementation of the ROM and not use pyMOR's generic `Model` classes ..

**Reply by ftalbrecht** [reply]
> I would not go through the trouble of extending the `Model` interface (and maybe even the `Operator` interface as well), just to save a few characters.

Sure, it was not about saving chars.

> Regarding runtime considerations: If your ROM is so fast that the Python loop matters, then you probably want a custom implementation of the ROM and not use pyMOR's generic `Model` classes ..

Agreed.
